connector/spanmetricsconnector/internal/metrics/metrics.go

```diff
@@ -0,0 +1,258 @@
+	if !ok {
+		agg := new(structure.Histogram[float64])
+		cfg := structure.NewConfig(
+			structure.WithMaxSize(m.MaxSize),
+		)
+		agg.Init(cfg)
+
+		h = &ExponentialHistogram{
+			Agg:        agg,
+			Attributes: attributes,
+			Exemplars:  pmetric.NewExemplarSlice(),
+		}
+		m.Metrics[key] = h
+	}
+
+	return h
+}
+
+func (m *ExponentialHistogramMetrics) BuildMetrics(
+	metric pmetric.Metric,
+	start pcommon.Timestamp,
+	temporality pmetric.AggregationTemporality,
+) {
+	metric.SetEmptyExponentialHistogram().SetAggregationTemporality(temporality)
+	dps := metric.ExponentialHistogram().DataPoints()
+	dps.EnsureCapacity(len(m.Metrics))
+	timestamp := pcommon.NewTimestampFromTime(time.Now())
+	for _, h := range m.Metrics {
+		dp := dps.AppendEmpty()
+		dp.SetStartTimestamp(start)
+		dp.SetTimestamp(timestamp)
+		expoHistToExponentialDataPoint(h.Agg, dp)
+		for i := 0; i < dp.Exemplars().Len(); i++ {
+			dp.Exemplars().At(i).SetTimestamp(timestamp)
+		}
+		h.Attributes.CopyTo(dp.Attributes())
+	}
+}
+
+// expoHistToExponentialDataPoint copies `lightstep/go-expohisto` structure.Histogram to
+// pmetric.ExponentialHistogramDataPoint
+func expoHistToExponentialDataPoint(agg *structure.Histogram[float64], dp pmetric.ExponentialHistogramDataPoint) {
+	dp.SetCount(agg.Count())
+	dp.SetSum(agg.Sum())
+	if agg.Count() != 0 {
+		dp.SetMin(agg.Min())
+		dp.SetMax(agg.Max())
+	}
+
+	dp.SetZeroCount(agg.ZeroCount())
+	dp.SetScale(agg.Scale())
+
+	for _, half := range []struct {
+		inFunc  func() *structure.Buckets
+		outFunc func() pmetric.ExponentialHistogramDataPointBuckets
+	}{
+		{agg.Positive, dp.Positive},
+		{agg.Negative, dp.Negative},
+	} {
+		in := half.inFunc()
+		out := half.outFunc()
+		out.SetOffset(in.Offset())
+		out.BucketCounts().EnsureCapacity(int(in.Len()))
+
+		for i := uint32(0); i < in.Len(); i++ {
+			out.BucketCounts().Append(in.At(i))
+		}
+	}
+}
+
+func (m *ExponentialHistogramMetrics) Reset(onlyExemplars bool) {
+	if onlyExemplars {
+		for _, h := range m.Metrics {
+			h.Exemplars = pmetric.NewExemplarSlice()
+		}
+		return
+	}
+
+	m.Metrics = make(map[Key]*ExponentialHistogram)
+}
+
+func (h *ExplicitHistogram) Observe(value float64) {
+	h.Sum += value
+	h.Count++
+
+	// Binary search to find the latencyMs bucket index.
+	index := sort.SearchFloat64s(h.Bounds, value)
+	h.BucketCounts[index]++
+}
+
+func (h *ExplicitHistogram) AddExemplar(traceID pcommon.TraceID, spanID pcommon.SpanID, value float64) {
+	e := h.Exemplars.AppendEmpty()
+	e.SetTraceID(traceID)
+	e.SetSpanID(spanID)
+	e.SetDoubleValue(value)
+}
+
+func (h *ExponentialHistogram) Observe(value float64) {
+	h.Agg.Update(value)
+}
+
+func (h *ExponentialHistogram) AddExemplar(traceID pcommon.TraceID, spanID pcommon.SpanID, value float64) {
+	e := h.Exemplars.AppendEmpty()
+	e.SetTraceID(traceID)
+	e.SetSpanID(spanID)
+	e.SetDoubleValue(value)
+}
+
+type Sum struct {
+	Attributes pcommon.Map
+	Count      uint64
+}
+
+type SumMetrics struct {
+	Metrics map[Key]*Sum
+}
```

Comment:
I liked the approach you took with the `GetOrCreate` and `Observe` methods employed in the histogram types above, because it also means we won't need to export these fields and, instead, rely on the `GetOrCreate` and `Observe` methods to mutate internal state. It also means a more consistent logic/methods between all the metric types, which can lead to future benefits.

What do you think?

Reply:
Do you mean reusing the same interfaces to handle sums that we use for histogram metrics and histograms? I was thinking of doing smth like that, but I guess it would add another layer of abstraction coz we would probably have to introduce an interface specific to metrics like sums and gauge and having just a sum metric now i am not sure whether it worth it.

Reply:
see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/19144/commits/d98cf22a4aa3c3410783fdc6278642e85a8db94b